dev/com.ibm.ws.session/src/com/ibm/ws/session/AbstractSessionData.java

```diff
@@ -0,0 +1,463 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.session;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Vector;
+import java.util.logging.Level;
+
+import javax.servlet.http.HttpServletRequest;
+
+import com.ibm.websphere.security.WebSphereRuntimePermission;
+import com.ibm.websphere.servlet.session.IBMApplicationSession;
+import com.ibm.ws.session.http.AbstractHttpSession;
+import com.ibm.ws.session.utils.LoggingUtil;
+import com.ibm.wsspi.servlet.session.IBMSessionExt;
+import com.ibm.wsspi.session.ISession;
+
+/**
+ * The WAS Http Session adaptation.  Extends the core HttpSessionImpl and adds WAS-specific function
+ * <p> 
+ * Since Servlet 6.0 : 
+ *      Common methods moved up from SessionData. 
+ */
+public abstract class AbstractSessionData extends AbstractHttpSession implements IBMSessionExt {
+
+    protected static final String SECURITY_PROP_NAME = "com_ibm_ejs_security_httpsession_info";
+
+    protected boolean _hasSecurityInfo = true; // used in
+    // getAttributeNames/getValueNames
+    // we have to assume true and we'll set to false
+    // in getValueNames if the sec prop isn't found
+
+    protected LinkedList lockList;
+    protected HashMap locks = null;
+
+    protected String pathInfoForAppSession = null;
+    String sipCookieInfo = null;
+    protected SessionContext _sessCtx;
+    protected String appName;
+    private boolean affinityEstablished = false;
+    protected IBMApplicationSession mIBMApplicationSession = null;
+
+    protected final static WebSphereRuntimePermission invalidateAllPerm = new WebSphereRuntimePermission("accessInvalidateAll");
+
+    private static final String methodClassName = "AbstractSessionData";
+
+    private final static int PUT_SESSION_VALUE = 0;
+    private final static int GET_SESSION_VALUE = 1;
+    private final static int REMOVE_SESSION_VALUE = 2;
+    private final static int SET_IBM_APPLICATION_SESSION = 3;
+
+    private final static String methodNames[] = { "putSessionValue", "getSessionValue", "removeSessionValue", "setIBMApplicationSession" };
+
```

Comment:
added constructor and associated super() from derived class. Moved _httpSessionFacade into common AbstractSessionData and provided a return method

Reply:
I was envisioning that:
```
_httpSessionFacade = returnFacade();
```
would still be here and you would have 
```
abstract protected HttpSessionFacade returnFacade();
```
defined on this class that the subclasses would implement.

I would be fine with it being called createFacade actually which makes a lot more sense.

Reply:
In reality can't most of the logic in the subclass constructors be consolidated to this constructor so that it mostly just needs to call super in the subclasses?  Remove the dual maintenance and puts the initialization of the fields in the class where they are defined.

Reply:
1.  abstract protected HttpSessionFacade returnFacade();
That would cause  subclass HttpSessionImpl to be an abstract class which can break the current flow.
Instead, I made a 
    protected void createFacade(AbstractHttpSessionFacade sessionFacade) 

 and call in from subclasses to set the _httpSessionFacade.
    
2. Consolidated most of the subclasses' constructor logic to AbstractSessionData.  Subclasses constructors need to call into new method createFace(AbstractHttpSessionFacade) to set the session facade.